File: java/src/test/java/io/github/bareboneslib/bareboneshll/HLLPlusPlusTest.java
package io.github.bareboneslib.bareboneshll;

import org.junit.jupiter.api.*;
import static org.junit.jupiter.api.Assertions.*;
import net.openhft.hashing.LongHashFunction;

public class HLLPlusPlusTest {
    LongHashFunction hash = LongHashFunction.xx();
    // ------------------------------------------------------------------ //
    //  Smoke tests                                                         //
    // ------------------------------------------------------------------ //

    @Test
    void constructor_defaultParams_doesNotThrow() {
        assertDoesNotThrow(() -> new HLLPlusPlus(12, 6));
    }

    @Test
    void constructor_minBoundary_doesNotThrow() {
        assertDoesNotThrow(() -> new HLLPlusPlus(4, 4));
    }

    @Test
    void constructor_maxBoundary_doesNotThrow() {
        assertDoesNotThrow(() -> new HLLPlusPlus(18, 6));
    }

    @Test
    void constructor_invalidP_throws() {
        assertThrows(IllegalArgumentException.class, () -> new HLLPlusPlus(3, 6));
        assertThrows(IllegalArgumentException.class, () -> new HLLPlusPlus(19, 6));
    }

    @Test
    void constructor_invalidR_throws() {
        assertThrows(IllegalArgumentException.class, () -> new HLLPlusPlus(12, 3));
        assertThrows(IllegalArgumentException.class, () -> new HLLPlusPlus(12, 7));
    }

    @Test
    void estimate_emptySketch_returnsZero() {
        HLLPlusPlus hll = new HLLPlusPlus(12, 6);
        assertEquals(0, hll.estimate());
    }

    @Test
    void estimate_singleElement_returnsOne() {
        HLLPlusPlus hll = new HLLPlusPlus(12, 6);
        hll.add(hash.hashLong(42L));
        assertEquals(1, hll.estimate());
    }

    @Test
    void estimate_duplicates_countedOnce() {
        HLLPlusPlus hll = new HLLPlusPlus(12, 6);
        for (int i = 0; i < 1000; i++) hll.add(hash.hashLong(42L));
        assertEquals(1, hll.estimate());
    }

    // ------------------------------------------------------------------ //
    //  Accuracy / error-rate tests                                         //
    // ------------------------------------------------------------------ //

    /**
     * HLL++ theoretical std error = 1.04 / sqrt(m), m = 2^p.
     * For p=12 that's ~1.6%. We allow 3x that as a generous bound.
     */
    @Test
    void accuracy_smallCardinality_withinErrorBound() {
        assertWithinError(12, 6, 1_000, 0.05);
    }

    @Test
    void accuracy_mediumCardinality_withinErrorBound() {
        assertWithinError(12, 6, 100_000, 0.05);
    }

    @Test
    void accuracy_largeCardinality_withinErrorBound() {
        assertWithinError(12, 6, 1_000_000, 0.05);
    }

    @Test
    void accuracy_highPrecision_tighterBound() {
        // p=16 → std error ~0.4%, allow 2%
        assertWithinError(16, 6, 100_000, 0.02);
    }

    private void assertWithinError(int p, int r, long n, double maxRelError) {
        HLLPlusPlus hll = new HLLPlusPlus(p, r);
        for (long i = 0; i < n; i++) hll.add(hash.hashLong(i));
        long est = hll.estimate();
        double relError = Math.abs(est - n) / (double) n;
        assertTrue(relError <= maxRelError,
                String.format("p=%d n=%d est=%d relError=%.4f > maxAllowed=%.4f",
                        p, n, est, relError, maxRelError));
    }

    // ------------------------------------------------------------------ //
    //  Serialization / deserialization roundtrip                           //
    // ------------------------------------------------------------------ //

    @Test
    void serdeRoundtrip_emptySketch_estimatePreserved() {
        HLLPlusPlus original = new HLLPlusPlus(12, 6);
        HLLPlusPlus restored = HLLPlusPlus.deserialize(original.serialize());
        assertEquals(original.estimate(), restored.estimate());
    }

    @Test
    void serdeRoundtrip_populatedSketch_estimatePreserved() {
        HLLPlusPlus original = new HLLPlusPlus(12, 6);
        for (long i = 0; i < 100_000; i++) original.add(hash.hashLong(i));

        byte[] bytes = original.serialize();
        HLLPlusPlus restored = HLLPlusPlus.deserialize(bytes);

        assertEquals(original.estimate(), restored.estimate());
    }

    @Test
    void serdeRoundtrip_sparseSketch_estimatePreserved() {
        // small cardinality stays in sparse representation
        HLLPlusPlus original = new HLLPlusPlus(12, 6);
        for (long i = 0; i < 100; i++) original.add(hash.hashLong(i));

        HLLPlusPlus restored = HLLPlusPlus.deserialize(original.serialize());
        assertEquals(original.estimate(), restored.estimate());
    }

    @Test
    void serialize_nullInput_throws() {
        assertThrows(Exception.class, () -> HLLPlusPlus.deserialize(null));
    }

    @Test
    void serialize_corruptedBytes_throws() {
        assertThrows(Exception.class,
                () -> HLLPlusPlus.deserialize(new byte[]{0x00, 0x01, 0x02}));
    }

    // ------------------------------------------------------------------ //
    //  Merge tests                                                         //
    // ------------------------------------------------------------------ //

    @Test
    void merge_disjointSets_estimateApproximatesUnion() {
        HLLPlusPlus a = new HLLPlusPlus(12, 6);
        HLLPlusPlus b = new HLLPlusPlus(12, 6);

        for (long i = 0;        i < 50_000; i++) a.add(hash.hashLong(i));
        for (long i = 50_000;   i < 100_000; i++) b.add(hash.hashLong(i));

        boolean ok = a.merge(b);
        assertTrue(ok);

        long est = a.estimate();
        double relError = Math.abs(est - 100_000) / 100_000.0;
        assertTrue(relError <= 0.05,
                String.format("merge estimate=%d relError=%.4f", est, relError));
    }

    @Test
    void merge_overlappingSets_doesNotDoubleCount() {
        HLLPlusPlus a = new HLLPlusPlus(12, 6);
        HLLPlusPlus b = new HLLPlusPlus(12, 6);

        // identical sets — estimate should stay ~N, not ~2N
        for (long i = 0; i < 50_000; i++) { a.add(hash.hashLong(i)); b.add(hash.hashLong(i)); }

        a.merge(b);
        long est = a.estimate();
        double relError = Math.abs(est - 50_000) / 50_000.0;
        assertTrue(relError <= 0.05,
                String.format("overlap merge estimate=%d relError=%.4f", est, relError));
    }

    @Test
    void merge_emptyIntoPopulated_estimateUnchanged() {
        HLLPlusPlus populated = new HLLPlusPlus(12, 6);
        HLLPlusPlus empty     = new HLLPlusPlus(12, 6);

        for (long i = 0; i < 50_000; i++) populated.add(hash.hashLong(i));
        long before = populated.estimate();

        populated.merge(empty);
        assertEquals(before, populated.estimate());
    }

    @Test
    void merge_incompatibleParams_returnsFalse() {
        HLLPlusPlus a = new HLLPlusPlus(12, 6);
        HLLPlusPlus b = new HLLPlusPlus(14, 6); // different p
        assertFalse(a.merge(b));
    }
}

File: cpp/test/HLLPlusPlusTest.cpp
#include <gtest/gtest.h>
#include <xxhash.h>
#include <iomanip> 
#include "../src/HLLPlusPlus.h"

// ---------------------------------------------------------------------------
// Helper — hash a 64-bit integer with xxHash before feeding to the sketch
// ---------------------------------------------------------------------------
static inline uint64_t h(uint64_t value) {
    return XXH64(&value, sizeof(value), 0);
}

static void addRange(HLLPlusPlus& hll, uint64_t from, uint64_t to) {
    for (uint64_t i = from; i < to; i++) hll.add(h(i));
}

static double relError(long est, long actual) {
    return std::abs(est - actual) / (double) actual;
}

// ===========================================================================
//  Smoke tests
// ===========================================================================

TEST(Smoke, ValidConstruction) {
    EXPECT_NO_THROW(HLLPlusPlus(12, 6));
    EXPECT_NO_THROW(HLLPlusPlus(4,  4));   // min boundary
    EXPECT_NO_THROW(HLLPlusPlus(18, 6));   // max boundary
}

TEST(Smoke, InvalidP_Throws) {
    EXPECT_THROW(HLLPlusPlus(3,  6), std::invalid_argument);
    EXPECT_THROW(HLLPlusPlus(19, 6), std::invalid_argument);
}

TEST(Smoke, InvalidR_Throws) {
    EXPECT_THROW(HLLPlusPlus(12, 3), std::invalid_argument);
    EXPECT_THROW(HLLPlusPlus(12, 7), std::invalid_argument);
}

TEST(Smoke, EmptySketch_EstimateIsZero) {
    HLLPlusPlus hll(12, 6);
    EXPECT_EQ(0, hll.estimate());
}

TEST(Smoke, SingleElement_EstimateIsOne) {
    HLLPlusPlus hll(12, 6);
    hll.add(h(42));
    EXPECT_EQ(1, hll.estimate());
}

TEST(Smoke, Duplicates_CountedOnce) {
    HLLPlusPlus hll(12, 6);
    for (int i = 0; i < 1000; i++) hll.add(h(42));
    EXPECT_EQ(1, hll.estimate());
}

// ===========================================================================
//  Accuracy / error-rate tests
// ===========================================================================

// HLL++ std error = 1.04 / sqrt(2^p)
// p=12 → ~1.6%  — we allow 5% (3x theoretical)
// p=16 → ~0.4%  — we allow 2%

TEST(Accuracy, SmallCardinality_p12) {
    HLLPlusPlus hll(12, 6);
    addRange(hll, 0, 1'000);
    EXPECT_LE(relError(hll.estimate(), 1'000), 0.05);
}

TEST(Accuracy, MediumCardinality_p12) {
    HLLPlusPlus hll(12, 6);
    addRange(hll, 0, 100'000);
    EXPECT_LE(relError(hll.estimate(), 100'000), 0.05);
}

TEST(Accuracy, LargeCardinality_p12) {
    HLLPlusPlus hll(12, 6);
    addRange(hll, 0, 1'000'000);
    EXPECT_LE(relError(hll.estimate(), 1'000'000), 0.05);
}

TEST(Accuracy, HighPrecision_p16) {
    HLLPlusPlus hll(16, 6);
    addRange(hll, 0, 100'000);

    long est      = hll.estimate();
    long actual   = 100'000;
    double rel    = relError(est, actual);

    EXPECT_LE(rel, 0.02)
        << "  actual:    " << actual    << "\n"
        << "  estimate:  " << est       << "\n"
        << "  abs error: " << std::abs(est - actual) << "\n"
        << "  rel error: " << std::fixed << std::setprecision(4) << rel * 100 << "%\n"
        << "  allowed:   2.00%";
}

// ===========================================================================
//  Serialization / deserialization roundtrip
// ===========================================================================

TEST(Serde, EmptySketch_RoundtripPreservesEstimate) {
    HLLPlusPlus original(12, 6);
    auto bytes = original.serialize();
    auto restored = HLLPlusPlus::deserialize(bytes);
    EXPECT_EQ(original.estimate(), restored.estimate());
}

TEST(Serde, PopulatedSketch_RoundtripPreservesEstimate) {
    HLLPlusPlus original(12, 6);
    addRange(original, 0, 100'000);

    auto bytes = original.serialize();
    auto restored = HLLPlusPlus::deserialize(bytes);
    EXPECT_EQ(original.estimate(), restored.estimate());
}

TEST(Serde, SparseSketch_RoundtripPreservesEstimate) {
    // small cardinality stays in sparse representation
    HLLPlusPlus original(12, 6);
    addRange(original, 0, 100);

    auto bytes = original.serialize();
    auto restored = HLLPlusPlus::deserialize(bytes);
    EXPECT_EQ(original.estimate(), restored.estimate());
}

TEST(Serde, CorruptedBytes_Throws) {
    std::vector<uint8_t> garbage = {0x00, 0x01, 0x02};
    EXPECT_THROW(HLLPlusPlus::deserialize(garbage), std::exception);
}

// ===========================================================================
//  Merge tests
// ===========================================================================

TEST(Merge, DisjointSets_ApproximatesUnion) {
    HLLPlusPlus a(12, 6), b(12, 6);
    addRange(a, 0,      50'000);
    addRange(b, 50'000, 100'000);

    EXPECT_TRUE(a.merge(b));
    EXPECT_LE(relError(a.estimate(), 100'000), 0.05);
}

TEST(Merge, OverlappingSets_NoDoubleCounting) {
    HLLPlusPlus a(12, 6), b(12, 6);
    addRange(a, 0, 50'000);
    addRange(b, 0, 50'000);   // identical

    EXPECT_TRUE(a.merge(b));
    EXPECT_LE(relError(a.estimate(), 50'000), 0.05);
}

TEST(Merge, EmptyIntoPopulated_EstimateUnchanged) {
    HLLPlusPlus populated(12, 6), empty(12, 6);
    addRange(populated, 0, 50'000);
    long before = populated.estimate();

    EXPECT_TRUE(populated.merge(empty));
    EXPECT_EQ(before, populated.estimate());
}

TEST(Merge, IncompatibleParams_ReturnsFalse) {
    HLLPlusPlus a(12, 6), b(14, 6);   // different p
    EXPECT_FALSE(a.merge(b));
}